enhancements/deprecated-bundles.md

```diff
@@ -25,60 +25,49 @@ status: implementable
-
-Versions that were removed from the index will not receive updates if subscriptions still exist for them on cluster. Ideally, appropriate communication from the operator maintainer will mean that cluster admins will upgrade to the latest supported version prior to cluster upgrade. In the event that the cluster is upgraded and `Subscriptions` to terminated bundles exist on cluster, the cluster admin can manually trigger an upgrade by either:
-
-- editing the `Subscription`'s `channel` field to start receiving updates from a different channel
-- creating a custom `CatalogSource` pointing to the previous OCP index, editing the `Subscription`'s `source` field to point to the new `CatalogSource` in order to receive updates.
-
-With the new work on skip range being steppable, the above scenario may be less likely to happen.
+We propose to add a command to `opm index` called `deprecate` which takes in a set of bundle images, an index, and all the same arguments as `opm index add`. The command modifies the index by marking all versions that come before the specified bundle image in the update graph (that is, until the tail of each channel) as deprecated as well. The deprecation of bundles from the update graph will not modify the update graph after the deprecated version (from the deprecated channel to the head of the channel).
 
 A new field will be added to the operator-registry database table `operatorbundle` called `deprecated` and the bundle images specified in the command will be marked as such.
 
-The OLM resolver will be modified with a new property type called `olm.deprecated` in order to filter out any options that have this property set and disallow installs of deprecated bundles. This also means appropriate status reporting in the subscription and installplan need to be added to OLM, and alerts will be fired if a deprecated version is installed.
+The OLM resolver will be modified with a new property type called `olm.deprecated` in order to filter out any options that have this property set and disallow installs of deprecated bundles either directly or during upgrade. This also means appropriate status reporting in the subscription and installplan need to be added to OLM, and alerts will be fired if a deprecated version is installed.
+
+**Note**: OLM cannot know whether an upgrade from a deprecated bundle to a non-deprecated bundle is supported / possible for a given operator. For example, given 1.0.0 -> 1.0.1 -> 1.0.2 -> 1.0.3, the team may have tested and implemented only sequential upgrades. Deprecating 1.0.1 from the update graph will mean installs on 1.0.0 will update directly to 1.0.2 which may not have been tested or implemented by the operator team.
```

Comment:
Isn't this in direct disagreement with the statement made earlier?

> The command modifies the index by marking all versions that come before the specified bundle image in the update graph (that is, until the tail of each channel) as deprecated as well.

I thought this intentionally did not attempt any graph stitching, but instead just deprecated any previous versions. And how does that work with respect to clusters that are on a version that has several versions ahead of it deprecated? They still need to exist to be stepped through, right? Or is that also explicitly disallowed and anyone using deprecation needs to provide an update edge directly to the latest supported version?

Reply:
since the resolver would filter all these deprecated versions out during resolution, it would update from the deprecated version directly to a version that is not deprecated right? My point is that this hop may not be supported or tested by the team - they may only have tested the path of stepping through intermediate versions.

Reply:
OLM is never going to install something that doesn't have a direct upgrade path defined by the index. So sure, it can filter things out, but it will also not add edges between previously disconnected nodes.

Reply:
it's not about adding edges.  Suppose someone already installed 1.0.0 before it was deprecated, and they have not upgraded.

Then the new catalog comes out with 1.0.0 and 1.0.1 deprecated, and the customer says "hey now i want to upgrade".  They will get sent directly to 1.0.2?  They won't be able to upgrade (because upgrading requires installing 1.0.1 first)?  The upgrade will go ahead and install 1.0.1 anyway (before going to 1.0.2)?

that's the behavior we need to define (and think about the implications of).  It's not entirely a new problem.  Today we more or less strand users who call into this situation (because 1.0.1 won't even exist in the catalog, so there will be no path to 1.0.2 at all, short of going back and installing an older catalog that includes 1.0.1).